sympy/functions/elementary/trigonometric.py

```diff
@@ -3029,8 +3029,13 @@ def eval(cls, y, x):
                     return -S.Pi/2
                 elif y.is_zero:
                     return S.NaN
-        if y.is_zero and x.is_extended_real and fuzzy_not(x.is_zero):
-            return S.Pi * (S.One - Heaviside(x))
+        if y.is_zero:
+            if x.is_extended_real and fuzzy_not(x.is_zero):
+                return S.Pi * (S.One - Heaviside(x))
+            if x.is_number:
+                return Piecewise((S.Pi, re(x) < 0),
+                                 (0, (re(x) > 0) |  Ne(im(x), 0)),
+                                 (S.NaN, True))
```

Comment:
Is the Nan True not implicit? I guess it's fine to be explicit anyway.

We are assuming here that x is complex so, given that the first condition would be already false can the second condition just be `Ne(x, 0)`?

Reply:
Well, there is the `re(x) == 0` case as well, and in this case we assume that we do not know what it is. A typical example may be `atan(sqrt(2)/(2*(sqrt(2)/2 + 1)))**2 + log(sqrt(2)/2 - sqrt(2)*I/2 - sqrt(2)*sqrt(I)*I/2 + sqrt(2)*sqrt(I)/2+sqrt(I))**2` which gives None for most of the assumptions tests. It may very well be zero, but we cannot really tell. Similarly, the problematic `sin(1)**2 + cos(1)**2-1` will go here as it cannot be decided (except that it happened to evaluate to -1e-134...). So I think that the most correct description is this. Earlier it was (not here, but in _eval_rewrite_as_atan), x < 0, x > 0, True, which neglected the complex case.

Maybe the nan, True is implicit. No idea. Seems like a bad thing to count on it though.

Reply:
From the docstring:
https://github.com/sympy/sympy/blob/12c50d4004174407a6afc9ab748ba553de42aa07/sympy/functions/elementary/piecewise.py#L69-L76
What this mean is in this case is that
1. if `re(x) < 0` is True the result is pi.
2. if `re(x) < 0` is False and `(re(x) > 0 | im(x) != 0)` is True then the result is 0
3. Otherwise it's a nan because of the 3rd rule but it would be even if the third rule wasn't there.

Given that the *combined* condition for the second case is `re(x) >= 0 & (re(x) > 0 | im(x) != 0)` is that not equivalent to `re(x) >= 0 & x != 0` which is what it would be if the second condition in the Piecewise was `Ne(x, 0)`?

Reply:
Ah, sorry. I did clearly not read carefully enough, I thought it said `Ne(im(x), 0)`. (There is a parallel discussion about Piecewise simplification in #17283, btw, where I put forward exactly this type of reasoning, cond & ~prev_cond...)